## Investigasi sampel data titanic berikut dengan cara :
1. Cek secara head, tail, sample, info lalu observasi apa yang bisa anda peroleh ?
2. Lakukan Statistical Summary dengan mengekstrak informasi yang didapat dari observasi anda ?
3. Cek apakah ada duplikat dan bagaimana handlenya ?
4. Cek apakah ada missing value, berapa persentasenya jika ada, dan bagaimana cara handlenya ?

## Import Libraries

In [44]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", None)

In [46]:
# load data from excel
df = pd.read_excel('titanic.xlsx')

# show the head / 5 first data
df.head()

survived                                             name     sex      age
0         1                    Allen, Miss. Elisabeth Walton  female  29.0000
1         1                   Allison, Master. Hudson Trevor    male   0.9167
2         0                     Allison, Miss. Helen Loraine  female   2.0000
3         0             Allison, Mr. Hudson Joshua Creighton    male  30.0000
4         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female  25.0000

In [47]:
# get data from bottom / last
df.tail()

survived                                     name     sex   age
495         1  Mallet, Mrs. Albert (Antoinette Magnin)  female  24.0
496         0        Mangiavacchi, Mr. Serafino Emilio    male   NaN
497         0               Matthews, Mr. William John    male  30.0
498         0                Maybery, Mr. Frank Hubert    male  40.0
499         0                McCrae, Mr. Arthur Gordon    male  32.0

In [50]:
#get 5 random data
df.sample(5)

survived                               name   sex   age
445         0        Hickman, Mr. Stanley George  male  21.0
221         0     Ostby, Mr. Engelhart Cornelius  male  65.0
22          1              Behr, Mr. Karl Howell  male  26.0
399         1      Drew, Master. Marshall Brines  male   8.0
47          1  Calderhead, Mr. Edward Pennington  male  42.0

# cek data missing and duplicated

In [53]:
# total duplicate from df
df.duplicated().sum()

1

In [55]:
# displays duplicate data
duplicate = df[df.duplicated(keep=False)]
print (duplicate)

     survived                            name     sex   age
104         1  Eustis, Miss. Elizabeth Mussey  female  54.0
349         1  Eustis, Miss. Elizabeth Mussey  female  54.0


In [57]:
# searching nul from dataframe
df.isnull().sum()

survived     0
name         0
sex          0
age         49
dtype: int64

In [165]:
#ppercentage Null Value from index
precentace = df["age"].isnull().sum()
index_df = len(df)
t_precen = ((precentace/index_df)*100)
t_precen

9.8

In [59]:
#all data null from dfdf
missing_data = df[df.isnull().any(axis=1)]
missing_data

survived                                               name     sex  age
15          0                                Baumann, Mr. John D    male  NaN
37          1      Bradley, Mr. George ("George Arthur Brayton")    male  NaN
40          0                          Brewe, Dr. Arthur Jackson    male  NaN
46          0                              Cairns, Mr. Alexander    male  NaN
59          1  Cassebeer, Mrs. Henry Arthur Jr (Eleanor Genev...  female  NaN
69          1             Chibnall, Mrs. (Edith Martha Bowerman)  female  NaN
70          0              Chisholm, Mr. Roderick Robert Crispin    male  NaN
74          0                        Clifford, Mr. George Quincy    male  NaN
80          0                          Crafton, Mr. John Bertram    male  NaN
106         0                                 Farthing, Mr. John    male  NaN
107         1               Flegenheim, Mrs. Alfred (Antoinette)  female  NaN
108         1                            Fleming, Miss. Margaret  female  NaN
118         0                        Franklin, Mr. Thomas Parham    male  NaN
121         1  Frauenthal, Mrs. Henry William (Clara Heinshei...  female  NaN
125         0                                   Fry, Mr. Richard    male  NaN
134         1       Goldenberg, Mrs. Samuel L (Edwiga Grabowska)  female  NaN
147         0                          Harrington, Mr. Charles H    male  NaN
152         1                        Hawksford, Mr. Walter James    male  NaN
157         0                        Hilliard, Mr. Herbert Henry    male  NaN
166         0                           Hoyt, Mr. William Fisher    male  NaN
176         1                  Kenyon, Mrs. Frederick R (Marion)  female  NaN
179         0                                 Klaber, Mr. Herman    male  NaN
184         0                                  Lewy, Mr. Ervin G    male  NaN
196         1                               Marechal, Mr. Pierre    male  NaN
204         1              Meyer, Mrs. Edgar Joseph (Leila Saks)  female  NaN
219         1                          Omont, Mr. Alfred Fernand    male  NaN
223         0                      Parr, Mr. William Henry Marsh    male  NaN
235         1                Rheims, Mr. George Alexander Lucien    male  NaN
237         0                                Robbins, Mr. Victor    male  NaN
241         0                              Rood, Mr. Hugh Roscoe    male  NaN
254         1                              Saalfeld, Mr. Adolphe    male  NaN
256         1                             Salomon, Mr. Abraham L    male  NaN
269         0                         Smith, Mr. Richard William    male  NaN
277         1     Spencer, Mrs. William Augustus (Marie Eugenie)  female  NaN
283         0                              Stewart, Mr. Albert A    male  NaN
293         1  Taylor, Mrs. Elmer Zebley (Juliet Cummins Wright)  female  NaN
297         1                     Thorne, Mrs. Gertrude Maybelle  female  NaN
318         0             Williams-Lambert, Mr. Fletcher Fellows    male  NaN
320         1                                  Woolner, Mr. Hugh    male  NaN
364         0                              Campbell, Mr. William    male  NaN
383         0  Corey, Mrs. Percy C (Mary Phyllis Elizabeth Mi...  female  NaN
385         0                     Cunningham, Mr. Alfred Fleming    male  NaN
411         0                   Frost, Mr. Anthony Wood "Archie"    male  NaN
470         1                                Keane, Miss. Nora A  female  NaN
474         0                               Knight, Mr. Robert J    male  NaN
478         0                              Lamb, Mr. John Joseph    male  NaN
484         1                         Leitch, Miss. Jessie Wills  female  NaN
492         0                                Malachard, Mr. Noel    male  NaN
496         0                  Mangiavacchi, Mr. Serafino Emilio    male  NaN

In [61]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   survived  500 non-null    int64  
 1   name      500 non-null    object 
 2   sex       500 non-null    object 
 3   age       451 non-null    float64
dtypes: float64(1), int64(1), object(2)
memory usage: 15.8+ KB


# Observations:
1. total index 500 dan colums just 4
2. "survived" binary (0,1), maybe 0 for death and 1 for live
3. "age" just numerical
4. data null / empty just in colums age maybe id card missing or stowaway
5. duplicate data is an index 104 and 349
6. the percentage of missing values ​​is 9.8 percent of the total index

In [64]:
df. describe()

survived         age
count  500.000000  451.000000
mean     0.540000   35.917775
std      0.498897   14.766454
min      0.000000    0.666700
25%      0.000000   24.000000
50%      1.000000   35.000000
75%      1.000000   47.000000
max      1.000000   80.000000

# handle missing value and null

In [112]:
data = df.copy()

In [131]:
# fill in the NA data in the age column with the average
data['age'].fillna(df['age'].mean(),inplace=True)  
data

C:\Users\USER\AppData\Local\Temp\ipykernel_2032\264504416.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data['age'].fillna(df['age'].mean(),inplace=True)


survived                                               name     sex  \
0           1                      Allen, Miss. Elisabeth Walton  female   
1           1                     Allison, Master. Hudson Trevor    male   
2           0                       Allison, Miss. Helen Loraine  female   
3           0               Allison, Mr. Hudson Joshua Creighton    male   
4           0    Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   
5           1                                Anderson, Mr. Harry    male   
6           1                  Andrews, Miss. Kornelia Theodosia  female   
7           0                             Andrews, Mr. Thomas Jr    male   
8           1      Appleton, Mrs. Edward Dale (Charlotte Lamson)  female   
9           0                            Artagaveytia, Mr. Ramon    male   
10          0                             Astor, Col. John Jacob    male   
11          1  Astor, Mrs. John Jacob (Madeleine Talmadge Force)  female   
12          1                      Aubart, Mme. Leontine Pauline  female   
13          1                       Barber, Miss. Ellen "Nellie"  female   
14          1               Barkworth, Mr. Algernon Henry Wilson    male   
15          0                                Baumann, Mr. John D    male   
16          0                           Baxter, Mr. Quigg Edmond    male   
17          1    Baxter, Mrs. James (Helene DeLaudeniere Chaput)  female   
18          1                              Bazzani, Miss. Albina  female   
19          0                               Beattie, Mr. Thomson    male   
20          1                      Beckwith, Mr. Richard Leonard    male   
21          1   Beckwith, Mrs. Richard Leonard (Sallie Monypeny)  female   
22          1                              Behr, Mr. Karl Howell    male   
23          1                              Bidois, Miss. Rosalie  female   
24          1                                  Bird, Miss. Ellen  female   
25          0                                Birnbaum, Mr. Jakob    male   
26          1                            Bishop, Mr. Dickinson H    male   
27          1            Bishop, Mrs. Dickinson H (Helen Walton)  female   
28          1                             Bissette, Miss. Amelia  female   
29          1          Bjornstrom-Steffansson, Mr. Mauritz Hakan    male   
30          0                       Blackwell, Mr. Stephen Weart    male   
31          1                                   Blank, Mr. Henry    male   
32          1                            Bonnell, Miss. Caroline  female   
33          1                           Bonnell, Miss. Elizabeth  female   
34          0                           Borebank, Mr. John James    male   
35          1                           Bowen, Miss. Grace Scott  female   
36          1                        Bowerman, Miss. Elsie Edith  female   
37          1      Bradley, Mr. George ("George Arthur Brayton")    male   
38          0                            Brady, Mr. John Bertram    male   
39          0                                 Brandeis, Mr. Emil    male   
40          0                          Brewe, Dr. Arthur Jackson    male   
41          1          Brown, Mrs. James Joseph (Margaret Tobin)  female   
42          1     Brown, Mrs. John Murray (Caroline Lane Lamson)  female   
43          1    Bucknell, Mrs. William Robert (Emma Eliza Ward)  female   
44          1                    Burns, Miss. Elizabeth Margaret  female   
45          0                  Butt, Major. Archibald Willingham    male   
46          0                              Cairns, Mr. Alexander    male   
47          1                  Calderhead, Mr. Edward Pennington    male   
48          1   Candee, Mrs. Edward (Helen Churchill Hungerford)  female   
49          1                 Cardeza, Mr. Thomas Drake Martinez    male   
50          1  Cardeza, Mrs. James Warburton Martinez (Charlo...  female   
51          0                           Carlsson, Mr.

In [120]:
# Delete duplicate data
data.drop_duplicates(inplace=True)

In [122]:
# check deleted duplicate data
data.duplicated().sum()

0

In [124]:
#reset index 
data = data.reset_index(drop=True)

In [126]:
data

survived                                               name     sex  \
0           1                      Allen, Miss. Elisabeth Walton  female   
1           1                     Allison, Master. Hudson Trevor    male   
2           0                       Allison, Miss. Helen Loraine  female   
3           0               Allison, Mr. Hudson Joshua Creighton    male   
4           0    Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   
5           1                                Anderson, Mr. Harry    male   
6           1                  Andrews, Miss. Kornelia Theodosia  female   
7           0                             Andrews, Mr. Thomas Jr    male   
8           1      Appleton, Mrs. Edward Dale (Charlotte Lamson)  female   
9           0                            Artagaveytia, Mr. Ramon    male   
10          0                             Astor, Col. John Jacob    male   
11          1  Astor, Mrs. John Jacob (Madeleine Talmadge Force)  female   
12          1                      Aubart, Mme. Leontine Pauline  female   
13          1                       Barber, Miss. Ellen "Nellie"  female   
14          1               Barkworth, Mr. Algernon Henry Wilson    male   
15          0                                Baumann, Mr. John D    male   
16          0                           Baxter, Mr. Quigg Edmond    male   
17          1    Baxter, Mrs. James (Helene DeLaudeniere Chaput)  female   
18          1                              Bazzani, Miss. Albina  female   
19          0                               Beattie, Mr. Thomson    male   
20          1                      Beckwith, Mr. Richard Leonard    male   
21          1   Beckwith, Mrs. Richard Leonard (Sallie Monypeny)  female   
22          1                              Behr, Mr. Karl Howell    male   
23          1                              Bidois, Miss. Rosalie  female   
24          1                                  Bird, Miss. Ellen  female   
25          0                                Birnbaum, Mr. Jakob    male   
26          1                            Bishop, Mr. Dickinson H    male   
27          1            Bishop, Mrs. Dickinson H (Helen Walton)  female   
28          1                             Bissette, Miss. Amelia  female   
29          1          Bjornstrom-Steffansson, Mr. Mauritz Hakan    male   
30          0                       Blackwell, Mr. Stephen Weart    male   
31          1                                   Blank, Mr. Henry    male   
32          1                            Bonnell, Miss. Caroline  female   
33          1                           Bonnell, Miss. Elizabeth  female   
34          0                           Borebank, Mr. John James    male   
35          1                           Bowen, Miss. Grace Scott  female   
36          1                        Bowerman, Miss. Elsie Edith  female   
37          1      Bradley, Mr. George ("George Arthur Brayton")    male   
38          0                            Brady, Mr. John Bertram    male   
39          0                                 Brandeis, Mr. Emil    male   
40          0                          Brewe, Dr. Arthur Jackson    male   
41          1          Brown, Mrs. James Joseph (Margaret Tobin)  female   
42          1     Brown, Mrs. John Murray (Caroline Lane Lamson)  female   
43          1    Bucknell, Mrs. William Robert (Emma Eliza Ward)  female   
44          1                    Burns, Miss. Elizabeth Margaret  female   
45          0                  Butt, Major. Archibald Willingham    male   
46          0                              Cairns, Mr. Alexander    male   
47          1                  Calderhead, Mr. Edward Pennington    male   
48          1   Candee, Mrs. Edward (Helen Churchill Hungerford)  female   
49          1                 Cardeza, Mr. Thomas Drake Martinez    male   
50          1  Cardeza, Mrs. James Warburton Martinez (Charlo...  female   
51          0                           Carlsson, Mr.

In [167]:
data.to_excel('Titanic_clean.xlsx', index=False)

# Note
After replacing the NA data with the average age of all passengers to simplify data processing and analysis, then deleting duplicate data and tidying up the index.

In [80]:
data.columns

Index(['survived', 'name', 'sex', 'age'], dtype='object')

In [81]:
data["survived"].value_counts()

survived
1    269
0    230
Name: count, dtype: int64

# Observation
1. the number of people who died was around **230 people** and live **269 people**

# Statistical Summary

In [145]:
# group column names based on type
# it will make our life easier onwards
categoricals = ['name','sex']

numericals = ['survived','age']

In [150]:
data[numericals].describe()

survived         age
count  499.000000  499.000000
mean     0.539078   35.881538
std      0.498971   14.013337
min      0.000000    0.666700
25%      0.000000   25.500000
50%      1.000000   35.917775
75%      1.000000   45.000000
max      1.000000   80.000000

In [153]:
data[categoricals].describe()

name   sex
count                             499   499
unique                            499     2
top     Allen, Miss. Elisabeth Walton  male
freq                                1   288

In [157]:
for col in categoricals:
  print(f"Value counts of {col} column")
  print(data[col].value_counts(), '\n')

Value counts of name column
name
Allen, Miss. Elisabeth Walton                                                         1
Angle, Mr. William A                                                                  1
Becker, Miss. Ruth Elizabeth                                                          1
Becker, Miss. Marion Louise                                                           1
Becker, Master. Richard F                                                             1
Beauchamp, Mr. Henry James                                                            1
Beane, Mrs. Edward (Ethel Clarke)                                                     1
Beane, Mr. Edward                                                                     1
Bateman, Rev. Robert James                                                            1
Banfield, Mr. Frederick James                                                         1
Ball, Mrs. (Ada E Hall)                                                               1

In [160]:
for col in data.columns:
  print(f"==== {col} ====")
  print(data[col].value_counts(), '\n')

==== survived ====
survived
1    269
0    230
Name: count, dtype: int64 

==== name ====
name
Allen, Miss. Elisabeth Walton                                                         1
Angle, Mr. William A                                                                  1
Becker, Miss. Ruth Elizabeth                                                          1
Becker, Miss. Marion Louise                                                           1
Becker, Master. Richard F                                                             1
Beauchamp, Mr. Henry James                                                            1
Beane, Mrs. Edward (Ethel Clarke)                                                     1
Beane, Mr. Edward                                                                     1
Bateman, Rev. Robert James                                                            1
Banfield, Mr. Frederick James                                                         1
Ball, Mrs. (Ada E Hall)   